commit f02316adb4baf4dceaab79cc0ef4c2acdb544f3e
Author: John Spray <jspray@redhat.com>
Date:   Thu Dec 14 06:24:18 2017 -0500

    mgr: enable inter-module calls
    
    This is being done by passing native CPython objects
    back and forth.  It's safe because sub-interpreters in CPython
    share memory allocation infrastructure and share the GIL.
    
    With a view to PEP554, we limit inter-interpreter calls
    to pickleable objects, so that this may be implemented
    using byte-arrays in future.
    
    This infrastructure should enable:
     - the dashboard to display the status of other modules, for
       example the set of progress indicators from `progress`
     - dashboard and restful to share an underlying long running
       job mechanism.
    
    Signed-off-by: John Spray <john.spray@redhat.com>

diff --git a/src/mgr/ActivePyModule.cc b/src/mgr/ActivePyModule.cc
index c89334e08a3..425b3f99c7f 100644
--- a/src/mgr/ActivePyModule.cc
+++ b/src/mgr/ActivePyModule.cc
@@ -100,6 +100,61 @@ void ActivePyModule::notify_clog(const LogEntry &log_entry)
   }
 }
 
+bool ActivePyModule::method_exists(const std::string &method) const
+{
+  Gil gil(py_module->pMyThreadState, true);
+
+  auto boundMethod = PyObject_GetAttrString(pClassInstance, method.c_str());
+  if (boundMethod == nullptr) {
+    return false;
+  } else {
+    Py_DECREF(boundMethod);
+    return true;
+  }
+}
+
+PyObject *ActivePyModule::dispatch_remote(
+    const std::string &method,
+    PyObject *args,
+    PyObject *kwargs,
+    std::string *err)
+{
+  assert(err != nullptr);
+
+  // Rather than serializing arguments, pass the CPython objects.
+  // Works because we happen to know that the subinterpreter
+  // implementation shares a GIL, allocator, deallocator and GC state, so
+  // it's okay to pass the objects between subinterpreters.
+  // But in future this might involve serialization to support a CSP-aware
+  // future Python interpreter a la PEP554
+
+  Gil gil(py_module->pMyThreadState, true);
+
+  // Fire the receiving method
+  auto boundMethod = PyObject_GetAttrString(pClassInstance, method.c_str());
+
+  // Caller should have done method_exists check first!
+  assert(boundMethod != nullptr);
+
+  dout(20) << "Calling " << py_module->get_name()
+           << "." << method << "..." << dendl;
+
+  auto remoteResult = PyObject_Call(boundMethod,
+      args, kwargs);
+  Py_DECREF(boundMethod);
+
+  if (remoteResult == nullptr) {
+    // Because the caller is in a different context, we can't let this
+    // exception bubble up, need to re-raise it from the caller's
+    // context later.
+    *err = handle_pyerror();
+  } else {
+    dout(20) << "Success calling '" << method << "'" << dendl;
+  }
+
+  return remoteResult;
+}
+
 
 
 int ActivePyModule::handle_command(
diff --git a/src/mgr/ActivePyModule.h b/src/mgr/ActivePyModule.h
index ebb1ba182be..6b5eeda616c 100644
--- a/src/mgr/ActivePyModule.h
+++ b/src/mgr/ActivePyModule.h
@@ -52,12 +52,21 @@ public:
   void notify(const std::string &notify_type, const std::string &notify_id);
   void notify_clog(const LogEntry &le);
 
+  bool method_exists(const std::string &method) const;
+
+  PyObject *dispatch_remote(
+      const std::string &method,
+      PyObject *args,
+      PyObject *kwargs,
+      std::string *err);
+
   int handle_command(
     const cmdmap_t &cmdmap,
     const bufferlist &inbuf,
     std::stringstream *ds,
     std::stringstream *ss);
 
+
   void set_health_checks(health_check_map_t&& c) {
     health_checks = std::move(c);
   }
